### PR TITLE
Add tests for PiEnc proofs

### LINKED ISSUE #109: Write tests for PiEncProof
A `PiEncProof` has six elements that must satisfy three different constraints. Some ideas for tests:

- Swap each of the proof elements for a different, random element (of the correct type). Proof shouldn't verify.
- Try making the proof with inconsistent secret inputs (e.g. `k` is not used to make `K`, `rho` is not the correct nonce for `K`). Proof either shouldn't build correctly or shouldn't verify.
- Try making the proof with consistent secret inputs that are out of range (e.g. `k` is actually used to make `K`, but it's larger than `2^ell`).
- Try verifying the proof with different setup or public parameters than were used to construct it.

Related issue: #46.

### PR DESCRIPTION
**Pull request by marsella**
Closes #109

It adds the tests described in the issue. I didn't think of any others to put in.

It also fixes a bug! The `plaintext_mask` should not be passed with the proof. If the verifier knows the mask for the plaintext, they can unmask it. Yay for testing.

### CODE REVIEW

**Review by amaloz — COMMENTED** [review]
At first glance this looks great. However, if we could wait until #144 is merged before merging this in, we could revisit the tests here given the new `RingPedersen` API.

**Comment by marsella** [comment]
I was thinking the same thing as I wrote all those `ZkSetupParameters` in. Happy to wait.

**Comment by amaloz** [comment]
> I was thinking the same thing as I wrote all those `ZkSetupParameters` in. Happy to wait.

#144 is now merged so we should be good to go using the new API!

**Comment by marsella** [comment]
> #144 is now merged so we should be good to go using the new API!

Thank you @amaloz! Rebased, this is ready for re-review.

File: src/zkp/pienc.rs
Comment (amaloz) [inline src/zkp/pienc.rs:475]: It might be good to save the ciphertext and reset things to avoid any future headaches if someone adds a test to the end and doesn't realize `input` has not been reset.

**Review by amaloz — APPROVED** [review]
Looks good! A couple of pretty minor comments.

File: src/zkp/pienc.rs
Comment (marsella) [reply src/zkp/pienc.rs:311]: I tried this and got interesting behavior. The proof expects an input in the range `±2^ell`, and it can guarantee that its in the range `±2^ell + epsilon`. When I tested it, I found mixed results for things in the range. I'm guessing it's kind of randomized for stuff in between -- like, it depends on the size of the challenge and the mask.

| input | result| 
| --- | --- |
| `2^ell` |  passes |
| `2^ell + 1` | passes |
| `2^(ell + epsilon/2)` | passes 80% of the time |
| `(2^ell+epsilon) - 1` | fails |
| `2^ell+epsilon` | fails |

I decided to add a test that `2^ell` always passes, and `2^ell+epsilon` always fails, since I think those two things both happen with high probability.

File: src/zkp/pienc.rs
Comment (marsella) [reply src/zkp/pienc.rs:475]: Rewrote this a bit to match the formatting you used in #151, so this should be an issue any more!

File: src/paillier.rs
Comment (marsella) [reply src/paillier.rs:61]: Yeah, that's much nicer.

File: src/zkp/pienc.rs
Comment (amaloz) [reply src/zkp/pienc.rs]: Interesting! I'll need to dig into the paper a bit to understand what the claim _really_ is, because Figure 14 makes it sound like there's a guarantee that `k \in 2^ell` at the top, but the range check says "The proof guarantees that k \in 2^{ell + epsilon}`, which it sounds like it doesn't necessarily...

Comment (marsella) [reply src/zkp/pienc.rs]: Yeah, I took a quick look at the proof of soundness but I don't know all the notation they're using, will have to look more carefully at some point.

**Pull request merged**
